storage/block.go

```diff
@@ -57,67 +60,111 @@ func NewDatabaseBlock(start time.Time, encoder m3db.Encoder, opts m3db.DatabaseO
 		opts:     opts,
 		start:    start,
 		encoder:  encoder,
+		stream:   nil,
 		ctxAlloc: ctxAlloc,
 		ctx:      ctxAlloc(),
 		closed:   false,
+		writable: true,
 	}
 }
 
 func (b *dbBlock) StartTime() time.Time {
 	return b.start
 }
 
+func (b *dbBlock) IsSealed() bool {
+	return !b.writable
+}
+
 func (b *dbBlock) Write(timestamp time.Time, value float64, unit xtime.Unit, annotation []byte) error {
 	if b.closed {
 		return errWriteToClosedBlock
 	}
+	if !b.writable {
+		return errWriteToSealedBlock
+	}
 	return b.encoder.Encode(m3db.Datapoint{Timestamp: timestamp, Value: value}, unit, annotation)
 }
 
 func (b *dbBlock) Stream(blocker m3db.Context) (m3db.SegmentReader, error) {
 	if b.closed {
 		return nil, errReadFromClosedBlock
 	}
-	s := b.encoder.Stream()
 	if blocker != nil {
 		b.ctx.DependsOn(blocker)
 	}
+	s := b.stream
+	if b.writable {
+		s = b.encoder.Stream()
+	}
 	return s, nil
 }
 
-// close closes internal context and returns encoder to pool.
-func (b *dbBlock) closeContextAndEncoder() {
+// close closes internal context and returns encoder and stream to pool.
+func (b *dbBlock) close() {
 	// If the context is nil (e.g., when it's just obtained from the pool),
 	// we return immediately.
 	if b.ctx == nil {
 		return
 	}
-	if encoder := b.encoder; encoder != nil {
-		b.ctx.RegisterCloser(encoder.Close)
+	if b.writable {
```

Comment:
Fair enough.  You could do something that I've done in other places successfully to avoid this type of big branching:

```go
cleanup := func() {
	b.ctx.Close()
	b.ctx = nil
	b.encoder = nil
	b.stream = nil
}
if b.writable {
	// If the block is not sealed, we need to close the encoder.
	if encoder := b.encoder; encoder != nil {
		b.ctx.RegisterCloser(encoder.Close)
	}
	cleanup()
	return
}
if stream := b.stream; stream != nil {
	b.ctx.RegisterCloser(func() {
		if bytesPool := b.opts.GetBytesPool(); bytesPool != nil {
			segment := stream.Segment()
			stream.Reset(m3db.Segment{})
			bytesPool.Put(segment.Head)
			bytesPool.Put(segment.Tail)
		}
		stream.Close()
	})
}
cleanup()
```